lib/msf/core/web_services/servlet_helper.rb

```diff
@@ -98,7 +98,7 @@ def sanitize_params(params, query_hash = {})
     if query_hash.key?('id')
       raise ArgumentError, ("'id' is not a valid query parameter. Please use /api/v1/<resource>/{ID} instead.")
     end
-    params.symbolize_keys.except(:captures, :splat)
+    params.symbolize_keys.except(:captures, :splat).to_h.symbolize_keys
```

Comment:
I believe this will fix the issue at hand, but I'm a little concerned about the semantic differences between ruby 2.7 and 3.0. I imagine this might impact other places in the codebase that don't have tests.

### 2.7

`params` is a `Sinatra::IndifferentHash` instance:
```
[15] pry(#<Msf::WebServices::JsonRpcApp>)> params.class
=> Sinatra::IndifferentHash
```

`symbolize_keys` is an active_support method:
```
[16] pry(#<Msf::WebServices::JsonRpcApp>)> params.method(:symbolize_keys)
=> #<Method: Sinatra::IndifferentHash(Hash)#symbolize_keys() /Users/adfoster/.rvm/gems/ruby-3.0.1/gems/activesupport-6.1.3.2/lib/active_support/core_ext/hash/keys.rb:27>
```

The chained `except` method is also active support:
```
[6] pry(#<Msf::WebServices::JsonRpcApp>)> params.method(:except)
=> #<Method: Sinatra::IndifferentHash(Hash)#except(*keys) /Users/adfoster/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activesupport-5.2.6/lib/active_support/core_ext/hash/except.rb:12>
```

`symbolize_keys` in Ruby 2.7 + Rails is actually noop, and the keys remain strings:
```
[7] pry(#<Msf::WebServices::JsonRpcApp>)> params.symbolize_keys
=> {"{\"jsonrpc\":\"2.0\",\"method\":\"health.check\",\"id\":1,\"params\":"=>
  [{"}"=>nil}],
 "version"=>"v1"}
```

### 3.0

`params` is a `Sinatra::IndifferentHash` instance:
```
[15] pry(#<Msf::WebServices::JsonRpcApp>)> params.class
=> Sinatra::IndifferentHash
```

`symbolize_keys` is an active_support method:
```
[20] pry(#<Msf::WebServices::JsonRpcApp>)> params.method(:symbolize_keys)

=> #<Method: Sinatra::IndifferentHash(Hash)#symbolize_keys() /Users/adfoster/.rvm/gems/ruby-3.0.1/gems/activesupport-6.1.3.2/lib/active_support/core_ext/hash/keys.rb:27>```

The chained `except` method is a pure ruby method implemented in C:
```
[23] pry(#<Msf::WebServices::JsonRpcApp>)> params.method(:except)
=> #<Method: Sinatra::IndifferentHash(Hash)#except(*)>
[24] pry(#<Msf::WebServices::JsonRpcApp>)> params.method(:except).source_location

=> nil
```

Which means the `except` method no longer returns a `Sinatra::IndifferentHash` instance, but just a plain hash with string keys:

```
[25] pry(#<Msf::WebServices::JsonRpcApp>)> params.symbolize_keys.except(:captures, :splat).class

=> Hash
[26] pry(#<Msf::WebServices::JsonRpcApp>)> params.symbolize_keys.except(:captures, :splat)

=> {"{\"jsonrpc\":\"2.0\",\"method\":\"health.check\",\"id\":1,\"params\":"=>
  [{"}"=>nil}],
 "version"=>"v1"}
```

Reply:
Running into the same problem here too, will need to dig deeper then :+1:

https://github.com/rapid7/metasploit-framework/blob/f8940ffc6f1408082018390a04d991da3ffd26e6/lib/msf/core/module_manager/reloading.rb#L23-L24

### 2.7

```
>> self.class
=> Msf::Modules::Auxiliary__Scanner__Smb__Smb_enumshares::MetasploitModule
>> self.class.parent
=> Msf::Modules::Auxiliary__Scanner__Smb__Smb_enumshares
>> self.class.method(:parent)
=> #<Method: Msf::Modules::Auxiliary__Scanner__Smb__Smb_enumshares::MetasploitModule.parent() /Users/user/.rvm/gems/ruby-2.7.2@metasploit-framework/gems/activesupport-5.2.4.4/lib/active_support/core_ext/module/introspection.rb:34>
>> 
```

### 3.0.1

```
[5] pry(#<Msf::ModuleManager>)> metasploit_class
=> Msf::Modules::Auxiliary__Scanner__Smb__Smb_enumshares::MetasploitModule
[6] pry(#<Msf::ModuleManager>)> metasploit_class.parent
NoMethodError: undefined method `parent' for Msf::Modules::Auxiliary__Scanner__Smb__Smb_enumshares::MetasploitModule:Class
Did you mean?  present?
from (pry):6:in `reload_module'
[7] pry(#<Msf::ModuleManager>)> metasploit_class.method(:parent)
NameError: undefined method `parent' for class `#<Class:Msf::Modules::Auxiliary__Scanner__Smb__Smb_enumshares::MetasploitModule>'
Did you mean?  print
from (pry):7:in `method'
```